### ISSUE TITLE
"failed to run html help compiler on index.hhp" with doxygen is 1.9.1

### ISSUE THREAD

**Issue by 1vanK** [doxygen/doxygen#8335]
Broken: doxygen-1.9.1.windows.x64.bin.zip and doxygen-1.9.1.windows.bin.zip

Doxygen 1.8.20 and 1.8.17 works perfectly

https://github.com/urho3d/Urho3D/runs/1706191458 > (win-vs, static, DX11) > Documentation

```
  Patching output file 373/480
  Patching output file 374/480
  Patching output file 375/480
  Patching output file 376/480
  Patching output file 377/480
  Patching output file 378/480
  Patching output file 379/480
  Patching output file 380/480
  Patching output file 381/480
  Patching output file 382/480
  Patching output file 383/480
  Patching output file 384/480
  Patching output file 385/480
  Patching output file 386/480
  Patching output file 387/480
  Patching output file 388/480
  Patching output file 389/480
CUSTOMBUILD : error : failed to run html help compiler on index.hhp [D:\a\Urho3D\Urho3D\build\ci\Docs\doc.vcxproj]
  Patching output file 390/480
  Patching output file 391/480
  Patching output file 392/480
  Patching output file 393/480
  Patching output file 394/480
  Patching output file 395/480
  Patching output file 396/480
  Patching output file 397/480
  Patching output file 398/480
  Patching output file 399/480
  Patching output file 400/480
  Patching output file 401/480
  Patching output file 402/480
  Patching output file 403/480
  Patching output file 404/480
  Patching output file 405/480
  Patching output file 406/480
  Patching output file 407/480
  Patching output file 408/480
  Patching output file 409/480
  Patching output file 410/480
  Patching output file 411/480
  Patching output file 412/480
  Patching output file 413/480
  Patching output file 414/480
  Patching output file 415/480
  Patching output file 416/480
  Patching output file 417/480
  Patching output file 418/480
  Patching output file 419/480
  Patching output file 420/480
  Patching output file 421/480
  Patching output file 422/480
  Patching output file 423/480
  Patching output file 424/480
  Patching output file 425/480
  Patching output file 426/480
  Patching output file 427/480
  Patching output file 428/480
  Patching output file 429/480
  Patching output file 430/480
  Patching output file 431/480
  Patching output file 432/480
  Patching output file 433/480
  Patching output file 434/480
  Patching output file 435/480
  Patching output file 436/480
  Patching output file 437/480
  Patching output file 438/480
  Patching output file 439/480
  Patching output file 440/480
  Patching output file 441/480
  Patching output file 442/480
  Patching output file 443/480
  Patching output file 444/480
  Patching output file 445/480
  Patching output file 446/480
  Patching output file 447/480
  Patching output file 448/480
  Patching output file 449/480
  Patching output file 450/480
  Patching output file 451/480
  Patching output file 452/480
  Patching output file 453/480
  Patching output file 454/480
  Patching output file 455/480
  Patching output file 456/480
  Patching output file 457/480
  Patching output file 458/480
  Patching output file 459/480
  Patching output file 460/480
  Patching output file 461/480
  Patching output file 462/480
  Patching output file 463/480
  Patching output file 464/480
  Patching output file 465/480
  Patching output file 466/480
  Patching output file 467/480
  Patching output file 468/480
  Patching output file 469/480
  Patching output file 470/480
  Patching output file 471/480
  Patching output file 472/480
  Patching output file 473/480
  Patching output file 474/480
  Patching output file 475/480
  Patching output file 476/480
  Patching output file 477/480
  Patching output file 478/480
  Patching output file 479/480
  Patching output file 480/480
  Running html help compiler...
  lookup cache used 37759/65536 hits=556515 misses=42609
  finished...
```

It looks like CHM file starts generating before all html pages have been generated.

https://github.com/doxygen/doxygen/blob/master/src/doxygen.cpp#L11955-L11963

I don't know how to reproduce the problem in a test project, but I can give instructions on how to compile the engine 

Generating the engine project:

```
set "PATH=c:\Programs\CMake_3_19\bin;c:\Programs\Doxygen_1_9_1\;c:\Programs\Graphviz\bin\;c:\Program Files (x86)\HTML Help Workshop\"

mkdir Build
cd Build
cmake.exe ../Urho3D -G "Visual Studio 16" -A Win32 -DURHO3D_OPENGL=1 -DURHO3D_ANGELSCRIPT=1 -DURHO3D_LUA=1 -DURHO3D_STATIC_RUNTIME=1 -DURHO3D_SAMPLES=1 -DURHO3D_TOOLS=0 -DURHO3D_DOCS=1 -DURHO3D_LIB_TYPE=STATIC -DURHO3D_DATABASE_ODBC=1 -DURHO3D_GENERATEBINDINGS=1
```

**Comment by 1vanK** [comment]
Despite this, CHM file is successfully created.

**Comment by albert-github** [comment]
> It looks like CHM file starts generating before all html pages have been generated.

This cannot be concluded from the given output. The output is written to the buffered  stream stdout and the message about `CUSTOMBUILD` is written to the, un-buffered, stderr. When the output is one way or another redirected the buffering might get involved here, to overcome this it is possible to use the `-b` option with doxygen (see also #8329).

I don't think the building of the CHM file is started before that the hhp file / html files were written. 

Strange is also that the chm file is generated correctly:
- were you able to inspect the chm file?
- What happens when you run just the hhc compiler on the index.hhp?

The project is quite large to run (I see in the log that it takes 20 minutes to build and another 10 minutes to generate the documentation).

**Comment by albert-github** [comment]
I had a look in one of the "artifact" files of a version where the documentation ran on Windows (i.e. with 1.8.20), but here I cannot find a generated chm file or and hhp file.
- did something change in the build procedure?
  -  more specific did, besides a newer version of doxygen, change anything in `rake`?

**Comment by 1vanK** [comment]
There is 2 builded versions of documentation: doxygen 1.9.1 and doxygen 1.8.20

Doxyfile: Docs\generated\Doxyfile
Result: Docs\html
Log: Docs\Win32\Debug\doc\doc.log

https://dropmefiles.com/Sioeu
https://dropmefiles.com/9NCQQ

**Comment by 1vanK** [comment]
The problem is not related to github actions, because I reproduce it on my PC. I just change used version of doxygne and have/not have error.

**Comment by albert-github** [comment]
I'm not a fan of external files as they might vanish (and these will vanish as well after 1 week) but here you don't have that much choice see the amount of data.

I do see that in both versions a chm file is generated.
I tried to recreate these chm files with the command `"C:\Program Files (x86)\HTML Help Workshop\hhc.exe" index.hhp` and te results are:

1.8.20
```
Microsoft HTML Help Compiler 4.74.8702

Compiling d:\speeltuin\issue_8335\1.8.20\Docs\html\Urho3D_Documentation_0.0.0.chm

Warning: Too many topics associated to the "Urho3D::Graphics" keyword.  Remaining 78 links truncated.
Warning: Too many topics associated to the "Urho3D::UIElement" keyword.  Remaining 57 links truncated.
Warning: Too many topics associated to the "Urho3D::Variant" keyword.  Remaining 47 links truncated.

Compile time: 1 minute, 7 seconds
15,133  Topics
342,877 Local links
1,242   Internet links
0       Graphics


Created ...\1.8.20\Docs\html\Urho3D_Documentation_0.0.0.chm, 78,259,016 bytes
Compression decreased file by 78,328,690 bytes.`
```

1.9.1
```
Microsoft HTML Help Compiler 4.74.8702

Compiling d:\speeltuin\issue_8335\1.9.1\Docs\html\Urho3D_Documentation_0.0.0.chm

HHC5013: Error: URL reference in the TOC cannot be resolved: "_spriter_data2_d_8cpp.html#a21f799c4a5e8e2845b093f8597708781".
HHC5013: Error: URL reference in the TOC cannot be resolved: "_spriter_data2_d_8cpp.html#a22d0d9973aad01759054d90dc9ee2ebb".
...
HHC5013: Error: URL reference in the TOC cannot be resolved: "_generated_enums_8cpp.html#a9706a237a58fd996ba76ef4f5d7beacd".
HHC5013: Error: URL reference in the TOC cannot be resolved: "_sprite_8cpp.html#aa993717fa9a1f8c650bd9836ff243806".
HHC5013: Error: URL reference in the TOC cannot be resolved: "_generated_enums_8cpp.html#a3c78ec9c43b40faa51beb9e7941074a4".
HHC5013: Error: URL reference in the TOC cannot be resolved: "_generated_enums_8cpp.html#ace9bad2757670af9f4d4374378eae751".
HHC5013: Error: URL reference in the TOC cannot be resolved: "_generated_enums_8cpp.html#a22331c7725606d12cd6d2049981696f4".
HHC5013: Error: URL reference in the TOC cannot be resolved: "_generated_enums_8cpp.html#a02a3873fe887e00c8f685e94fa9f5736".
HHC5013: Error: URL reference in the TOC cannot be resolved: "_raycast_vehicle_8cpp.html#a246ee8768cbf68deb0cb995da877117a".
HHC5013: Error: URL reference in the TOC cannot be resolved: "_material_8cpp.html#a1ebaad3d037077f6cff7286fef5bb861".
Warning: Too many topics associated to the "Urho3D::Graphics" keyword.  Remaining 78 links truncated.
Warning: Too many topics associated to the "Urho3D::UIElement" keyword.  Remaining 57 links truncated.
Warning: Too many topics associated to the "Urho3D::Variant" keyword.  Remaining 47 links truncated.


The following files were not compiled:
```

so this explains the error message.
We have to see why these `HHC5013: Error: URL reference in the TOC cannot be resolved: ` appear.
I searched for `a246ee8768cbf68deb0cb995da877117a` in the 1.8.20 and 1.91 html directories and for 1.8.20 there were no results and for 1.9.1 I got:
```
index.hhc:      <LI><OBJECT type="text/sitemap"><param name="Name" value="wheelElementNames"><param name="Local" value="_raycast_vehicle_8cpp.html#a246ee8768cbf68deb0cb995da877117a"><param name="ImageNumber" value="11"></OBJECT>
```
but in neither version the file `_raycast_vehicle_8cpp.html` is present nor is the name `wheelElementBames` present.

Searching for `wheelElementNames` I found in the sources:
```
Source/Urho3D/Physics/RaycastVehicle.cpp:static const StringVector wheelElementNames =
```
Which matches the filename and also the variable name.
- What directly can be seen that this is a `static` variable and these are by default not shown in the documentation.
- most likely cause is that some internal doxygen structure changed and that the writing of the `index.hhc` has not been adjusted to this.

This has to be investigated further.

**Comment by albert-github** [comment]
To facilitate error searching very much:
- could you please post the used `Doxyfile`, no need to put it on `dropmefiles` as it it small enough to add (as attachment) to a comment (and it will be persistent here).

**Comment by albert-github** [comment]
I found the `Doxyfile` as `Docs\generated\Doxyfile` in the files on "dropmefiles" and with a cloned version of the repository I was able to reproduce the problem. I'm trying to minimize the problem.

**Comment by albert-github** [comment]
I've been able to reduce the problem to a very minimal example:
```

namespace Urho3D
{

/// something fie
int fie()
{
    return 1;
}

/// something inner NS
namespace AttributeMetadata
{
    /// something inner var
    static const int j = 1;
}

}
```
and the settings:
```
INPUT                  = Serializable.h
GENERATE_HTMLHELP      = YES
HHC_LOCATION           = "C:/Program Files (x86)/HTML Help Workshop/hhc.exe"
SEARCHENGINE           = NO
GENERATE_LATEX         = NO
QUIET=YES
```

Problem seems to be that the outer namespace is not documented and thus the namespace is not seen (when I remember correctly some changes have been made here).
When adding e.g. `/// something NS` as documentation the error disappears as the namespace is shown correctly.

Example: [example.tar.gz](https://github.com/doxygen/doxygen/files/5821178/example.tar.gz)


It is not sufficient to just document the `namespace Urho3D` as at that moment there are still 3 messages left (regarding SpriterData2D.cpp):
```
HC5013: Error: URL reference in the TOC cannot be resolved: "_spriter_data2_d_8cpp.html#a21f799c4a5e8e2845b093f8597708781".
HHC5013: Error: URL reference in the TOC cannot be resolved: "_spriter_data2_d_8cpp.html#a22d0d9973aad01759054d90dc9ee2ebb".
HHC5013: Error: URL reference in the TOC cannot be resolved: "_spriter_data2_d_8cpp.html#ad5f04a52442abcddcba0bf814fdb9d03".
```
Documenting also the inner `namespace Spriter` here looks like to solve this problem.

**Comment by albert-github** [comment]
The problem has been introduce in 
```
Commit: e847d5dc322740b0097f56fc19311c8c74a44a55 [e847d5d]
Date: Monday, October 19, 2020 10:18:07 AM
issue #8103: C++ Table of content, namespace list does not contains namespace without class
```
the commit before it (ccca51f80ce6a4c380545800f6ea8b5ae27c8724 ) was still OK.
Also the commit
```
Commit: 33b0f4d25dff25b0e50d62eff68155106e88d58d [33b0f4d]
Date: Monday, October 19, 2020 4:03:07 PM
issue #8103: C++ Table of content, namespace list does not contains namespace without class (part 2)
```
didn't fix the problem.

**Comment by albert-github** [comment]
I did some more investigations and the problem can also be seen when we disable the `GENERATE_HTMLHELP` and enable `GENERATE_TREEVIEW`:

![Treeview](https://user-images.githubusercontent.com/5223533/104760130-36564600-5761-11eb-97c1-b007ad4b8e13.png)

when clicking on `fie` we get a page not found!

Problem looks like to be in the function `static void writeNamespaceTreeElement` in `index.cpp` or in one of the routines calling this routine / the used lists for these calling routines.

**Comment by 1vanK** [comment]
Thanks for your help

**Comment by albert-github** [comment]
I did a quick test on the repository and it loos to me that with this fix the problem is solved.
(I still think that in the project the mentioned namespaces should be documented, but that is an issue for the project).

**Comment by 1vanK** [comment]
Thanks!

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.2.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
